pkg/ddl/notifier/store.go

```diff
@@ -106,34 +126,85 @@ func (t *tableStore) DeleteAndCommit(
 	return errors.Trace(err)
 }
 
-func (t *tableStore) List(ctx context.Context, se *sess.Session) ([]*schemaChange, error) {
-	sql := fmt.Sprintf(`
-		SELECT
-			ddl_job_id,
-			sub_job_id,
-			schema_change,
-			processed_by_flag
-		FROM %s.%s ORDER BY ddl_job_id, sub_job_id`,
-		t.db, t.table)
-	rows, err := se.Execute(ctx, sql, "ddl_notifier")
+// List implements Store interface.
+func (t *tableStore) List(ctx context.Context, se *sess.Session) (ListResult, CloseFn) {
+	return &listResult{
+		ctx: ctx,
+		se:  se,
+		sqlTemplate: fmt.Sprintf(`
+			SELECT
+				ddl_job_id,
+				sub_job_id,
+				schema_change,
+				processed_by_flag
+			FROM %s.%s
+			WHERE (ddl_job_id, sub_job_id) > (%%?, %%?)
+			ORDER BY ddl_job_id, sub_job_id
+			LIMIT %%?`,
+			t.db, t.table),
+		// DDL job ID are always positive, so we can use 0 as the initial value.
+		maxReturnedDDLJobID: 0,
+		maxReturnedSubJobID: 0,
+	}, se.Rollback
+}
+
+type listResult struct {
+	ctx                 context.Context
+	se                  *sess.Session
+	sqlTemplate         string
+	maxReturnedDDLJobID int64
+	maxReturnedSubJobID int64
+}
+
+// Read implements ListResult interface.
+func (r *listResult) Read(changes []*SchemaChange) (int, error) {
+	if r.maxReturnedDDLJobID == 0 && r.maxReturnedSubJobID == 0 {
+		err := r.se.Begin(r.ctx)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+	}
+
+	rows, err := r.se.Execute(
+		r.ctx, r.sqlTemplate, "ddl_notifier",
+		r.maxReturnedDDLJobID, r.maxReturnedSubJobID, len(changes),
+	)
 	if err != nil {
-		return nil, err
+		return 0, errors.Trace(err)
 	}
-	ret := make([]*schemaChange, 0, len(rows))
-	for _, row := range rows {
-		event := SchemaChangeEvent{}
-		err = json.Unmarshal(row.GetBytes(2), &event)
+
+	if err = r.unmarshalSchemaChanges(rows, changes); err != nil {
+		return 0, errors.Trace(err)
+	}
+	return len(rows), nil
+}
+
+func (r *listResult) unmarshalSchemaChanges(rows []chunk.Row, changes []*SchemaChange) error {
+	for i, row := range rows {
+		if changes[i] == nil {
+			changes[i] = new(SchemaChange)
+		}
+		if changes[i].event == nil {
+			changes[i].event = new(SchemaChangeEvent)
+		}
+		if changes[i].event.inner == nil {
+			changes[i].event.inner = new(jsonSchemaChangeEvent)
+		}
```

Comment:
How about:
```go
// Add constructor function
func NewSchemaChange() *SchemaChange {
    return &SchemaChange{
        event: &SchemaChangeEvent{
            inner: &jsonSchemaChangeEvent{},
        },
    }
}

// Replace existing code
if changes[i] == nil {
    changes[i] = NewSchemaChange()
}
```

Reply:
Or do you mean there is a case that `changes[i]` is not nil, but the event or the inner is nil?

Reply:
just to improve robustness, I don't want to assume caller uses some kind of `SchemaChange`.

I'll test the OOM problem soon